Abstract class for scheduling and ShortestTaskFirst

talvarez0507 opened the pull request:
This PR changes FindSchedule into an abstract BaseTaskScheduler class which future algorithms will extend. It also updates the servlet to handle and use the algorithm type based on the request from the client. I also included my scheduling algorithm ShortestTaskFirst and its JUnit tests.

talvarez0507 committed:
add BaseTaskScheduler class

talvarez0507 committed:
add javadoc comment

talvarez0507 committed:
change to support new abstract class

talvarez0507 committed:
create ShortestTaskFirst.java

talvarez0507 committed:
delete no longer used file

talvarez0507 committed:
add JUnit tests for ShortestTaskFirst

hollyyuqizheng commented on portfolio/src/test/java/com/google/sps/ShortestTaskFirstTest.java:
```diff
@@ -0,0 +1,171 @@
+
+  // Makes sure that no errors are thrown if working hours is longer than a
+  // 24 hour period.
+  @Test
```
I was thinking that, for all the time validity checking, eg. is the user trying to schedule for a date that is before "today", is the user's working hours correct, will all happen on the client side, so ideally all the time input to any Java methods are all valid. So this test might end up not being too necessary, because the user's working hours will not be over 3 days, per checked on the client side already.

talvarez0507 replied:
The client side code can always be changed by the user so this an unlikely but possible case. This test was just to ensure no errors are thrown but maybe we should add a check somewhere in CalendarGroup?

hollyyuqizheng replied:
I see. I think this check will happen on the client side instead of in the calendar event group class, eg. I have already added some validity checking on the UI side of things, so I think the input to the scheduling algorithm should be in good shape.

hollyyuqizheng commented on portfolio/src/test/java/com/google/sps/ShortestTaskFirstTest.java:
```diff
@@ -0,0 +1,171 @@
+  }
+
+  // Makes sure that no errors are thrown for events outside working hours.
+  @Test
```
Maybe we could add another test for some events that lie partially outside the working hours. For example, for a 9-5 working hour, an event could be from 8:30 to 9:30, and the test would make sure that the first scheduled task is not at 9 or something.

talvarez0507 replied:
Done. Broke this test into two new ones, events partially/totally outside working hours.

hollyyuqizheng reviewed (changes requested):
Some suggestions on additional tests.

talvarez0507 committed:
add more granular tests

Two new tests: one that has an event partially outside of the working hours and one that is totally outside of the working hours.

hollyyuqizheng commented on portfolio/src/main/java/com/google/sps/data/ShortestTaskFirst.java:
```diff
@@ -36,7 +32,7 @@
    * based on the tasks that were able to be scheduled.
    */
-  public static Collection<ScheduledTask> shortestTaskFirst(
+  public Collection<ScheduledTask> schedule(
```
Just a reminder for us that calling the getEmptyTimeRanges will look a bit different since it's in another class now

talvarez0507 replied:
I fixed it now rather than later to prevent us from forgetting.

raulcruise commented on portfolio/src/test/java/com/google/sps/ShortestTaskFirstTest.java:
```diff
@@ -0,0 +1,187 @@
+
+  private final TaskPriority LOW_PRIORITY = new TaskPriority(1);
+
+  // Makes sure we return an empty list in the case where no tasks are passed.
```
Genuinely wondering, are we supposed to use Javadoc style for unit tests as well?

talvarez0507 replied:
Good question, I'll ask in our next pod sync.

jamesnam-google replied:
Yes please use javadoc style in test methods.

talvarez0507 replied:
Done.

talvarez0507 committed:
update to CalendarGroup.java implementation

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/BaseTaskScheduler.java:
```diff
@@ -0,0 +1,32 @@
+import java.time.Instant;
+import java.util.Collection;
+
+// TODO: Add test coverage for scheduling algorithms
```
Can we have username for the TODO.

talvarez0507 replied:
Removed the TODO. That was an old TODO for myself but I included the tests already in this PR.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/BaseTaskScheduler.java:
```diff
@@ -0,0 +1,32 @@
+
+// TODO: Add test coverage for scheduling algorithms
+public abstract class BaseTaskScheduler {
+  public enum SchedulingAlgorithmType {
```
Consider making this enum its own file, so it can be referenced easily in the code-base.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/BaseTaskScheduler.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.Collection;
+
+// TODO: Add test coverage for scheduling algorithms
+public abstract class BaseTaskScheduler {
```
I think this class can be package-private (i.e., take out the public keyword).

talvarez0507 replied:
I don't think I can do that because I use this class in the schedule servlet which is in a different package.

jamesnam-google replied:
I think you don't want to reference BaseTaskScheduler outside this package. What you can do is create an interface that is public, with the two methods. Then you can have the task scheduler classes implement that interface, and reference the interface elsewhere rather than this abstract class.

talvarez0507 replied:
In hindsight, some of the stuff I'm doing in this servlet makes it impossible to have schedulers be package-private. As a result I created a class called ServletUtils  to 1. clean up the servlet a bit 2. allow the schedulers to be package-private. All I did was copy over the existing helper methods and put them into that class.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/BaseTaskScheduler.java:
```diff
@@ -0,0 +1,32 @@
+  }
+
+  /** This is the abstract method that all algorithm classes should implement. */
+  public abstract Collection<ScheduledTask> schedule(
```
I think it can be package-private as well.

talvarez0507 replied:
Similar to my other comment, the servlet uses this method so I could not make it private.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/BaseTaskScheduler.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.Collection;
+
+// TODO: Add test coverage for scheduling algorithms
+public abstract class BaseTaskScheduler {
```
I think there should be an abstract method that returns the algorithm type.

abstract SchedulingAlgorithmType getSchedulingAlgorithmType();

And classes that implement the abstract class can override that method.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/ScheduledTask.java:
```diff
@@ -24,4 +28,13 @@ public Task getTask() {
+    return other instanceof ScheduledTask && equals(this, (ScheduledTask) other);
+  }
+
+  public static boolean equals(ScheduledTask a, ScheduledTask b) {
```
Is this method necessary? Each time we call ScheduledTask.equals(a, b), we can just call a.equals(b).

talvarez0507 replied:
This code came almost directly from a week 5 walk-through. This method is mainly to simplify the logic for the other .equals(). If I remove this then instead of `&& equals(this, (ScheduledTask) other)` in the first .equals(), I would need something like:
```
return other instanceof ScheduledTask
        && this.task.equals(((ScheduledTask) b).task)
        && this.startTime.equals(((ScheduledTask) b).startTime);
```
Do you recommend I leave this method in or remove it?

jamesnam-google replied:
You can leave it here if you want.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/ShortestTaskFirst.java:
```diff
@@ -36,17 +32,14 @@
-      Collection<Task> tasks,
-      Instant workHoursStartTime,
-      Instant workHoursEndTime) {
+  public Collection<ScheduledTask> schedule(Collection<CalendarEvent> events,
```
Consider declaring one param per line.

ex:
...schedule(
    Collection...
    Collection...

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -40,13 +41,32 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     Collection<CalendarEvent> events = collectEventsFromJsonArray(eventsArray);
     Collection<Task> tasks = collectTasksFromJsonArray(tasksArray);
+
+    // These are null for initialization purposes. They will either be set or
```
I don't think the program should throw an error in case of error. The servlet can return some error response, like 400 (Bad Request). And this servlet should complete gracefully.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -40,13 +41,32 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
+
+    // This will take the string representation of the algorithm that is passed
+    // in and match it to something in the enum under BaseTaskScheduler.
+    try {
```
Consider using a helper method that returns an optional containing SchedulingAlgorithmType.

ex:

Optional<SchedulingAlgorithmType> schedulingAlgorithmTypeOptional =
    getSchedulingAlgorithmTypeOptional(...);
if (!schedulingAlgorithmTypeOptional.isPresent()) {
  // Return a 400 response to the client and return the method.
}

Optional<BaseTaskScheduler> algorithmOptional = getAlgorithmOptional(...);
if (!algorithmOptional.isPresent()) {
  // Same thing here
}

...

rgcase replied:
I think we can skip the `schedulingAlgorithmType` variable too since it's only used here, and like James said just write a `getAlgorithmOptional(...)` method that takes in the String.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/test/java/com/google/sps/ShortestTaskFirstTest.java:
```diff
@@ -0,0 +1,187 @@
+  private static Instant END_OF_DAY = Instant.parse("2020-06-25T23:59:59Z");
+  private static Instant THREE_DAYS_LATER = Instant.parse("2020-06-28T00:00:00Z");
+
+  private static Duration TIME_15_MINUTES = Duration.ofSeconds(900);
```
Consider replacing 900 with 15 * 60 for readability. Here and below.

jamesnam-google replied:
Also, I think the Duration variables can be named DURATION_* instead of TIME_*. That way we can easily distinguish duration variables from instant variables.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/test/java/com/google/sps/ShortestTaskFirstTest.java:
```diff
@@ -0,0 +1,187 @@
+    Assert.assertEquals(expected, actual);
+  }
+
+  // Makes sure no errors are thrown if we have more tasks than we can possibly
```
You can consider including this info in the method name.

ex:
tooManyTasksToSchedule_shouldNotThrowError() {

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/ShortestTaskFirst.java:
```diff
@@ -21,13 +21,9 @@
-// TODO: Add test coverage for scheduling algorithms
-
-public class FindSchedule {
+public class ShortestTaskFirst extends BaseTaskScheduler {
```
Consider using suffix to indicate that this is a task scheduler.

ex:
ShortestTaskFirstScheduler

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/test/java/com/google/sps/ShortestTaskFirstTest.java:
```diff
@@ -0,0 +1,187 @@
+  // Makes sure we return an empty list in the case where no tasks are passed.
+  @Test
+  public void noTasksScheduled() {
+    ShortestTaskFirst algorithm = new ShortestTaskFirst();
```
Consider renaming the variable to shortestTaskFirstScheduler. I think algorithm is a vague name.

talvarez0507 replied:
Done.

talvarez0507 committed:
remove TODO, add getSchedulingAlgorithmType()

talvarez0507 committed:
update ScheduleServlet.java

Create new methods that handle the algorithm setting and implement Optional

talvarez0507 committed:
rename class, add getSchedulingAlgorithmType()

talvarez0507 committed:
extract enum into new file

talvarez0507 committed:
update test file name

talvarez0507 committed:
implement change requests

Change Duration constants to start with "DURATION_" instead of "TIME_". Reflect class name changes like ShortestTaskFirst to ShortestTaskFirstScheduler. Create more descriptive name for a test. Change calculation inside Duration.ofSeconds() to be minutes*60 for readability. Change 'algorithm' variable to shortestTaskFirstScheduler.

talvarez0507 committed:
add lint check fixes

talvarez0507 committed:
change illegal modifier

talvarez0507 committed:
change illegal modifier

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/BaseTaskScheduler.java:
```diff
@@ -0,0 +1,30 @@
+
+public abstract class BaseTaskScheduler {
+  /** This is the scheduling method that all algorithm classes should implement. */
+  public abstract Collection<ScheduledTask> schedule(Collection<CalendarEvent> events,
```
Consider declaring one param per line.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -91,4 +117,35 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
+      case "SHORTEST_TASK_FIRST":
+        return Optional.of(SchedulingAlgorithmType.SHORTEST_TASK_FIRST);
+    }
+    return null;
```
I think you want to return Optional.empty() instead of null.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -91,4 +117,35 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
+      case SHORTEST_TASK_FIRST:
+        return Optional.of(new ShortestTaskFirstScheduler());
+    }
+    return null;
```
Return Optional.empty()

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -91,4 +117,35 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
+    return null;
+  }
+
+  private static Optional<BaseTaskScheduler> getAlgorithmOptional(
```
Consider renaming to getTaskSchedulerOptional

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -40,27 +43,50 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
+    if (!schedulingAlgorithmTypeOptional.isPresent()) {
+      response.sendError(
+          HttpServletResponse.SC_BAD_REQUEST,
+          "The request by the client was syntactically incorrect.");
```
Consider clarifying that the algorithm type could not be determined.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -40,27 +43,50 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
+      return;
+    }
+
+    Optional<BaseTaskScheduler> algorithmOptional =
```
Consider renaming to taskSchedulerOptional. And as I suggested above, this variable can be of type Optional<TaskScheduler> where TaskScheduler is a new interface.

talvarez0507 replied:
Done.

jamesnam-google commented on portfolio/src/main/java/com/google/sps/servlets/ScheduleServlet.java:
```diff
@@ -40,27 +43,50 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
+    if (!algorithmOptional.isPresent()) {
+      response.sendError(
+          HttpServletResponse.SC_BAD_REQUEST,
+          "The request by the client was syntactically incorrect.");
```
Consider clarifying that the task scheduler could not be determined.

talvarez0507 replied:
Done.

talvarez0507 committed:
change to one param per line

talvarez0507 committed:
clarify algorithm error response

talvarez0507 committed:
fix incorrect string

talvarez0507 committed:
update return statements to not be null

talvarez0507 committed:
update to new ServletUtil methods

talvarez0507 committed:
change to package-private

talvarez0507 committed:
add ServletUtils.java

talvarez0507 committed:
update and rename to TaskScheduler.java

jamesnam-google commented on portfolio/src/main/java/com/google/sps/data/ServletUtils.java:
```diff
@@ -0,0 +1,92 @@
+    return tasks;
+  }
+
+  public static Optional<SchedulingAlgorithmType> getSchedulingAlgorithmTypeOptional(
```
I think this method and the one below can be in a separate file (similar to this class), like SchedulingAlgorithmReference.java. I usually try to avoid naming classes Util or Utils, because they are a bit ambiguous. But I'm fine with keeping this class named as ServletUtils (or maybe ServletHelper).

talvarez0507 replied:
Done.

jamesnam-google reviewed (approved):
Approving the change with a comment.

talvarez0507 committed:
rename file to ServletHelper and move methods

talvarez0507 committed:
create SchedulingAlgorithmReference.java

talvarez0507 committed:
reflect class name changes
